source/array.lua

```diff
@@ -0,0 +1,186 @@
+    Attributions and copyright licensing by Mozilla Contributors is licensed under CC-BY-SA 2.5
+
+    These are heuristcs tested across 300+KLOC of JS translated to Lua
+    It isn't perfect, but it works for dozens of packages in the React, Jest, and GraphQL ecosystem
+    In those dozens of packages, hundreds and hundreds of tests were ported to Lua and passing
+    Additionally, the packages were integrated into a 500+KLOC 60fps product and shipped to 150+M global users
+    Reimplemented by Matt Hargett, 2023
+
+]]
+--!strict
+local exports = {}
+export type Array<T> = { [number]: T }
+type Object = { [string]: any }
+exports.indexOf = function<T>(haystack: Array<T>, needle: T, startIndex_: number?): number
+  local length = #haystack
+  local startIndex = if startIndex_ == nil
+    then 1
+    else if startIndex_ < 1 then math.max(1, length - math.abs(startIndex_)) else startIndex_
+
+  for i = startIndex, length, 1 do
+    if haystack[i] == needle then
+      return i
+    end
+  end
+
+  -- we maintain the JS not found value, as it's often checked for explicitly and as < 0
+  return -1
+end
+
+-- we pulled a few refinement ideas from this stackoverflow article, but found:
+-- 1. no single one answer worked enough of the time in terms of transliterated JS expectations
+-- 2. most had very poor accuracy versus performance tradeoffs
+-- https://stackoverflow.com/questions/7526223/how-do-i-know-if-a-table-is-an-array/20958869#20958869
+exports.isArray = function(val: any): boolean
+  if type(val) ~= "table" then
+    return false
+  end
+
+  if next(val) == nil then
+    -- it's table with nothing in it, which we express is an array
+    -- this works 99% of the time for transliterated Lua
+    return true
+  end
+
+  local tableLength = #val
+  if tableLength == 0 then
+    -- getting past the preceding clause says the table isn't an empty iterable
+    -- if the length of the table is reported as 0, that means it has non-numeric indices
+    return false
+  end
+
+  -- the slow part, verifying each index is a positive, whole number. a Lua VM built-in would be nice.
+  for key, _ in pairs(val) do
+    if type(key) ~= "number" then
+      return false
+    end
+    if key < 1 then
+      -- Lua arrays start at 1, a 0 index means it's not a pure Lua array
+      return false
+    end
+    -- Lua TODO: would math.floor be faster? needs a benchmark
+    if (key % 1) ~= 0 then
+      -- if the number key isn't a whole number, it's not a pure Lua array
+      return false
+    end
+
+    if key > tableLength then
+      -- if we get a numeric key larger than the length operator reports, this isn't a contiguous array
+      return false
+    end
+
+    if key ~= tableLength then
+      -- if we're not at the end of a contiguous array, the value in the index slot should be non-nil
+      if nil == val[key + 1] then
+        return false
+      end
+    end
+  end
```

Comment:
I think you can replace all of this with a single check like this:
```suggestion
  local expectedIndex = 1
  for key, _ in val do
    if key ~= expectedIndex then
      return false
    end
    expectedIndex += 1
  end
```

Reply:
I like your approach better, but I want to have tests passing en masse before trying to optimize. This was very very tricky to get "right" acrosss React, Jest, GraphQL, etc the first time.